### DISCUSSION TITLE
from annotations to training & audio augmentations

### DISCUSSION THREAD

**Discussion by radekosmulski** [earthspecies/project#14] [General]
The 40 files I manually annotated ended up translating to 2200+ annotations! I am not sure how that is possible since the annotation process seemed to go so quickly. 

The files are compatible with Raven and I am reading them into Python. I constructed a Dataset around these files and am training a model on them right now. The results look promising. Since we have gone the full distance here, from annotations to training a deep learning model, maybe this is something we will find a use for, for this complete pipeline, in one of our partnerships. 

Other than that, I explored data augmentation. Unfortunately, to no avail. Frequency shift, time dilation, adding noise - none of this results in improvements in training. The reason for this might be that we are starting with quite a lot of data already and probably the model learns to distinguish among the 264 bird species based on very fine characteristics (I am not sure about this, just a theory at this point). I am however exploring other techniques that can help, currently focusing on PCEN (per channel energy normalization) and also looking into hopefully better ways to construct the spectrogram (foregoing resizing, modifying the frequency spectrum we are looking at, etc).

**Comment by aza** [comment]
2200 annotations!! Impressive.

Do you have a screenshot of the Electron app you made?

**Reply by radekosmulski** [reply]
Here you are 🙂 

![image](https://user-images.githubusercontent.com/2444926/88668821-680c1f80-d0e3-11ea-8e69-6a41f6bde61a.png)

**Reply by radekosmulski** [reply]
Continuing to research and experiment with various methods for the birdcall competition. PCEN turned out to not improve results, neither did resizing the spectrograms in a method that in theory should be better, should provide more information (instead of creating spectrograms with 128 mel bands and resizing to 224x, I anticipated spectrograms created with 224 mel band without resizing should work much better - this turned out to not be the case). The couple of next things I have on the agenda are training with label smoothing (really hopeful for that, both as the labels are weak and also we want the predictions of the models to be calibrated as best as possible) and with synthetically created examples from multiple recordings.

**Reply by bs** [reply]
- Are what's worked/hasn't worked being documented or conveyed to the Kaggle community?
- Could you plop this app on esp.dev? (the web version, not the electron app)

**Reply by radekosmulski** [reply]
Re sharing with the Kaggle community and beyond - that is the plan, in the process of drafting a write up
On the web application - I only have the electron version as it can access local files, haven't thought of converting it to a web version, haven't really thought about what it would require to do so, thinking that to convert this to a web app might be quite involved

**Reply by bs** [reply]
Ah, no worries if it's involved. Can revisit if we think easy of access would be useful.

Curious: Is file access a problem? I know Electron exposes os apis, but they aren't required, are they? (See https://audiomass.co/ as an example of a webapp that executes everything locally, including file access, after the initial fetch.)

**Reply by radekosmulski** [reply]
I am not really sure to be honest how this could work as a standalone web app. I mean this not in a sense 'I don't think it can work' just that I lack the understanding what it would take to make it happen.

The electron app uses os apis and I think all that could be implemented via dragging files onto the browser (like with audiomass) and saving files by clicking a button to download the generated annotation file, but not sure how much work that would require inside the app itself and to add these capabilities to the web page. Here I get all of this for free, I also can resave annotations whenever they change without user input.

I have never been too much of a frontend person but based on my understanding I think all of this can be done, I also don't think it would be all that involved (might be wrong on that one) and I anticipate it would take quite a bit of tweaking, probably on the order of a couple of days of work (buuut again, I might be horribly wrong on that one, maybe it is a quick business of changing a couple of things, its all HTML and javascript after all, but genuinely do not know how to port this to a browser, I for instance have never extensively processed an uploaded file in the frontend, always would do it in the backend, or at least as much as I could 😉)

**Reply by radekosmulski** [reply]
Sorry, thought about this some more, and I don't think it can be that hard 🙂 Just would need to add a button here and there.

**Comment by radekosmulski** [comment]
as for update - no new tangible outputs, continuing to experiment / bring work on the birdcall competition to closure, preparing for road map discussion, about to attend the Decoding Communication conference
